File: matlab/utils/load_and_validate_events.m
function events = load_and_validate_events( detected_events_file )
%LOAD_AND_VALIDATE_EVENTS Load and validate HFO event metadata from a .mat file.
%
%  Loads a .mat file containing HFO event metadata and validates that the
%  required fields ('start_idx', 'end_idx', 'chan_idx') are present. Also
%  ensures these fields are in integer format.
%
%  Inputs:
%    detected_events_file - String, path to a .mat file containing a struct with HFO events.
%                  The file must contain a struct with fields:
%                     - start_idx : Vector of start indices for each event
%                     - end_idx   : Vector of end indices for each event
%                     - chan_idx  : Vector of channel indices for each event
%
%  Outputs:
%    events - Struct containing validated and standardized HFO event fields.
%
%   Written by Shi-Bei (Ashley) Tan (tashibei@umich.edu) in 2025
%   Copyright © 2025 by Shi-Bei (Ashley) Tan is licensed under CC BY-NC 4.0

assert( isfile(detected_events_file), 'Events file not found: %s', detected_events_file );
events = load( detected_events_file );

required_fields = { 'start_idx', 'end_idx', 'chan_idx' };
for f = required_fields
    assert( isfield(events, f{1}), 'Events missing field: %s', f{1} );
end

events = validateAndConvertIntFields( events, {'start_idx', 'end_idx', 'chan_idx'} );

end

File: python/utils/load_and_validate_events.py
import os
import scipy.io as sio
import numpy as np
from utils.load_and_validate_eeg import validate_and_convert_int_fields


def load_and_validate_events(events_file):
    """
    Load and validate HFO event metadata from a .mat file.

    Args:
        events_file (str): Path to .mat file containing HFO events. The file
                           must contain fields:
                             - start_idx : event start indices
                             - end_idx   : event end indices
                             - chan_idx  : channel indices

    Returns:
        dict: Validated events with integer fields and 0-based indexing
    """

    if not os.path.isfile(events_file):
        raise FileNotFoundError(f"Events file not found: {events_file}")

    mat_contents = sio.loadmat(events_file)
    events = {k: v for k, v in mat_contents.items() if not k.startswith("__")}

    # Verify required fields
    required_fields = ['start_idx', 'end_idx', 'chan_idx']
    for f in required_fields:
        if f not in events:
            raise ValueError(f"Events missing field: {f}")

    # Convert to integer arrays
    events = validate_and_convert_int_fields(events, required_fields)

    # Adjust for 0-based indexing (MATLAB → Python)
    events['start_idx'] = events['start_idx'] - 1
    events['end_idx'] = events['end_idx'] - 1
    events['chan_idx'] = events['chan_idx'] - 1

    return events